File: server-go/internal/agentclient/convert.go
package agentclient

import (
	"errors"
	"log/slog"
	"strings"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/structpb"

	agentv1 "dovideo/server/gen/agent/v1"
	"dovideo/server/internal/common"
	"dovideo/server/internal/model"
)

// MapError translates a gRPC status into the error taxonomy (contract §2), so callers branch on
// meaning (permanent / budget / not ready / transient) and never on transport codes.
func MapError(err error) error {
	if err == nil {
		return nil
	}
	st, ok := status.FromError(err)
	if !ok {
		return common.Transient("agent call failed", err)
	}
	msg := st.Message()
	switch st.Code() {
	case codes.InvalidArgument:
		return &common.Error{Kind: common.KindInvalidArgument, Msg: msg, Cause: err}
	case codes.FailedPrecondition:
		return &common.Error{Kind: common.KindNotReady, Msg: common.NotReadyMessage, Cause: err}
	case codes.ResourceExhausted:
		var ae *agentError
		if errors.As(err, &ae) && ae.reason == "BUDGET_EXHAUSTED" {
			return common.BudgetExhausted(msg, err)
		}
		return common.Transient("agent overloaded: "+msg, err)
	case codes.Unauthenticated:
		// A token mismatch is a deployment error, not a request error: keep the task retryable so it
		// succeeds once the configuration is fixed, but make it loud.
		slog.Error("agent_rejected_internal_token", "err", msg)
		return common.Transient("agent rejected the internal token", err)
	}
	if strings.TrimSpace(msg) == "" {
		msg = "agent service " + st.Code().String()
	}
	return common.Transient(msg, err)
}

// agentError carries the x-agent-error trailer alongside the status error.
type agentError struct {
	error
	reason string
}

func (e *agentError) Unwrap() error { return e.error }

// GRPCStatus keeps status.FromError / status.Code working on the wrapped error.
func (e *agentError) GRPCStatus() *status.Status {
	st, _ := status.FromError(e.error)
	return st
}

// ---- enums ----

// ModeToProto maps a mode; unknown or empty modes become GENERAL.
func ModeToProto(m model.AnalysisMode) agentv1.AnalysisMode {
	switch m.Resolve() {
	case model.ModeLearning:
		return agentv1.AnalysisMode_LEARNING
	case model.ModeReview:
		return agentv1.AnalysisMode_REVIEW
	case model.ModeCreation:
		return agentv1.AnalysisMode_CREATION
	}
	return agentv1.AnalysisMode_GENERAL
}

// ModeFromProto maps a proto enum to its name; UNSPECIFIED is treated as GENERAL.
func ModeFromProto(m agentv1.AnalysisMode) model.AnalysisMode {
	switch m {
	case agentv1.AnalysisMode_LEARNING:
		return model.ModeLearning
	case agentv1.AnalysisMode_REVIEW:
		return model.ModeReview
	case agentv1.AnalysisMode_CREATION:
		return model.ModeCreation
	}
	return model.ModeGeneral
}

// ---- messages -> client JSON ----

func nonEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

// TaskStatusFromProto maps {state,result,message}; empty strings become null.
func TaskStatusFromProto(s *agentv1.TaskStatus) model.TaskStatus {
	state := model.State(s.GetState())
	if state == "" {
		state = model.StateCompleted
	}
	return model.TaskStatus{State: state, Result: nonEmpty(s.GetResult()), Message: nonEmpty(s.GetMessage())}
}

// EvidenceHit is the client JSON of one evidence search hit.
type EvidenceHit struct {
	StartMs    int64    `json:"startMs"`
	EndMs      int64    `json:"endMs"`
	Source     string   `json:"source"`
	Snippet    string   `json:"snippet"`
	Transcript string   `json:"transcript"`
	OcrTexts   []string `json:"ocrTexts"`
}

// HitsFromProto always returns a non-nil slice so the client receives [] rather than null.
func HitsFromProto(hits []*agentv1.VideoEvidenceHit) []EvidenceHit {
	out := make([]EvidenceHit, 0, len(hits))
	for _, h := range hits {
		ocr := h.GetOcrTexts()
		if ocr == nil {
			ocr = []string{}
		}
		out = append(out, EvidenceHit{StartMs: h.GetStartMs(), EndMs: h.GetEndMs(), Source: h.GetSource(),
			Snippet: h.GetSnippet(), Transcript: h.GetTranscript(), OcrTexts: ocr})
	}
	return out
}

// Plan is the client JSON of an agent plan.
type Plan struct {
	UnderstoodGoal string   `json:"understoodGoal"`
	Tasks          []string `json:"tasks"`
}

func PlanFromProto(p *agentv1.AgentPlan) *Plan {
	tasks := p.GetTasks()
	if tasks == nil {
		tasks = []string{}
	}
	return &Plan{UnderstoodGoal: p.GetUnderstoodGoal(), Tasks: tasks}
}

// StructToJSON passes a free-form Struct through as a JSON object; an empty struct means "none".
func StructToJSON(s *structpb.Struct) map[string]any {
	if s == nil || len(s.GetFields()) == 0 {
		return nil
	}
	return s.AsMap()
}

// FeedbackFromProto maps stored feedback to client JSON: unset optional fields become null.
func FeedbackFromProto(f *agentv1.AgentFeedback) model.AgentFeedback {
	id := f.GetMediaId()
	goal := f.GetGoal()
	mode := string(ModeFromProto(f.GetMode()))
	tasks := f.GetCorrectedTasks()
	if tasks == nil {
		tasks = []string{}
	}
	out := model.AgentFeedback{
		MediaID: &id, Goal: &goal, Mode: &mode,
		ErrorType: f.ErrorType, Comment: f.Comment, CorrectedGoal: f.CorrectedGoal, CorrectedTasks: tasks,
		EvidenceTimestamp: f.EvidenceTimestamp, EvidenceAccepted: f.EvidenceAccepted, CreatedAt: nonEmpty(f.GetCreatedAt()),
	}
	if f.Rating != nil {
		r := int(*f.Rating)
		out.Rating = &r
	}
	return out
}

// ---- requests ----

// FeedbackToProto maps validated user feedback; nil pointers stay unset.
func FeedbackToProto(f model.AgentFeedback) *agentv1.AgentFeedback {
	out := &agentv1.AgentFeedback{
		ErrorType: f.ErrorType, Comment: f.Comment, CorrectedGoal: f.CorrectedGoal, CorrectedTasks: f.CorrectedTasks,
		EvidenceTimestamp: f.EvidenceTimestamp, EvidenceAccepted: f.EvidenceAccepted,
	}
	if f.MediaID != nil {
		out.MediaId = *f.MediaID
	}
	if f.Goal != nil {
		out.Goal = *f.Goal
	}
	if f.Mode != nil {
		out.Mode = ModeToProto(model.ModeFromNullable(*f.Mode))
	}
	if f.Rating != nil {
		r := int32(*f.Rating)
		out.Rating = &r
	}
	if f.CreatedAt != nil {
		out.CreatedAt = *f.CreatedAt
	}
	return out
}

// SeedToProto converts the context-build telemetry.
func SeedToProto(s *TelemetrySeed) *agentv1.TelemetrySeed {
	if s == nil {
		return nil
	}
	out := &agentv1.TelemetrySeed{TraceId: s.TraceID, Stages: map[string]*agentv1.StageTiming{}, Counters: map[string]int64{}}
	for k, v := range s.Stages {
		out.Stages[k] = &agentv1.StageTiming{DurationMs: v.DurationMs, Success: v.Success}
	}
	for k, v := range s.Counters {
		out.Counters[k] = v
	}
	return out
}

File: agent-py/app/rpc/convert.py
"""The single place where protobuf messages and domain models meet.

Conventions: proto3 scalars cannot be null, so an empty string / zero means "absent" for plain fields;
fields declared ``optional`` in the proto map to ``None`` via HasField; ANALYSIS_MODE_UNSPECIFIED means
GENERAL; unknown enum numbers are rejected as INVALID_ARGUMENT.
"""
from datetime import datetime

from google.protobuf import struct_pb2

from ..errors import InvalidArgumentError
from ..gen.agent.v1 import agent_pb2 as pb
from ..models.dto import (
    AgentFeedback,
    AgentPlan,
    AgentState,
    AnalysisMode,
    AnalysisResult,
    CriticResult,
    Evidence,
    RouteDecision,
    Section,
    TaskStatus,
    VideoEvidenceHit,
    format_instant,
)

_FEEDBACK_OPTIONAL = ("rating", "error_type", "comment", "corrected_goal", "evidence_timestamp",
                      "evidence_accepted")


# ---------- enums ----------
def mode_from_proto(value: int) -> AnalysisMode:
    if value == pb.ANALYSIS_MODE_UNSPECIFIED:
        return AnalysisMode.GENERAL
    try:
        return AnalysisMode(pb.AnalysisMode.Name(value))
    except ValueError:
        raise InvalidArgumentError(f"不支持的分析模式: {value}") from None


def mode_to_proto(mode: AnalysisMode | str | None) -> int:
    if mode is None:
        return pb.GENERAL
    return pb.AnalysisMode.Value(mode.value if isinstance(mode, AnalysisMode) else AnalysisMode(mode).value)


def media_id(value: int, field: str = "media_id") -> int:
    if value <= 0:
        raise InvalidArgumentError(f"{field} is required")
    return value


# ---------- agent state ----------
def plan_to_proto(plan: AgentPlan) -> pb.AgentPlan:
    return pb.AgentPlan(understood_goal=plan.understood_goal, tasks=plan.tasks)


def plan_from_proto(msg: pb.AgentPlan) -> AgentPlan:
    return AgentPlan(understood_goal=msg.understood_goal, tasks=list(msg.tasks))


def result_to_proto(result: AnalysisResult) -> pb.AnalysisResult:
    return pb.AnalysisResult(
        title=result.title, conclusions=result.conclusions,
        evidence=[pb.Evidence(timestamp_ms=e.timestamp_ms, source=e.source, content=e.content, claim=e.claim)
                  for e in result.evidence],
        suggestions=result.suggestions,
        sections=[pb.Section(key=s.key, title=s.title, items=s.items) for s in result.sections])


def result_from_proto(msg: pb.AnalysisResult) -> AnalysisResult:
    return AnalysisResult(
        title=msg.title, conclusions=list(msg.conclusions),
        evidence=[Evidence(timestamp_ms=e.timestamp_ms, source=e.source, content=e.content, claim=e.claim)
                  for e in msg.evidence],
        suggestions=list(msg.suggestions),
        sections=[Section(key=s.key, title=s.title, items=list(s.items)) for s in msg.sections])


def critique_to_proto(critique: CriticResult) -> pb.CriticResult:
    return pb.CriticResult(passed=critique.passed, feedback=critique.feedback,
                           missing_requirements=critique.missing_requirements,
                           unsupported_claims=critique.unsupported_claims,
                           required_timestamps=critique.required_timestamps)


def critique_from_proto(msg: pb.CriticResult) -> CriticResult:
    return CriticResult(passed=msg.passed, feedback=list(msg.feedback),
                        missing_requirements=list(msg.missing_requirements),
                        unsupported_claims=list(msg.unsupported_claims),
                        required_timestamps=list(msg.required_timestamps))


def state_to_proto(state: AgentState) -> pb.AgentState:
    msg = pb.AgentState(goal=state.goal, round=state.round)
    if state.plan is not None:
        msg.plan.CopyFrom(plan_to_proto(state.plan))
    if state.result is not None:
        msg.result.CopyFrom(result_to_proto(state.result))
    if state.critique is not None:
        msg.critique.CopyFrom(critique_to_proto(state.critique))
    return msg


def state_from_proto(msg: pb.AgentState) -> AgentState:
    return AgentState(
        goal=msg.goal, round=msg.round,
        plan=plan_from_proto(msg.plan) if msg.HasField("plan") else None,
        result=result_from_proto(msg.result) if msg.HasField("result") else None,
        critique=critique_from_proto(msg.critique) if msg.HasField("critique") else None)


def task_status_to_proto(status: TaskStatus) -> pb.TaskStatus:
    return pb.TaskStatus(state=status.state.value, result=status.result or "", message=status.message or "")


# ---------- retrieval / routing ----------
def hit_to_proto(hit: VideoEvidenceHit) -> pb.VideoEvidenceHit:
    return pb.VideoEvidenceHit(start_ms=hit.start_ms, end_ms=hit.end_ms, source=hit.source,
                               snippet=hit.snippet, transcript=hit.transcript, ocr_texts=hit.ocr_texts)


def route_to_proto(decision: RouteDecision) -> pb.RouteDecision:
    return pb.RouteDecision(mode=mode_to_proto(decision.mode), reason=decision.reason)


# ---------- feedback ----------
def feedback_from_proto(msg: pb.AgentFeedback) -> AgentFeedback:
    values = {name: getattr(msg, name) if msg.HasField(name) else None for name in _FEEDBACK_OPTIONAL}
    created_at = None
    if msg.created_at:
        try:
            created_at = datetime.fromisoformat(msg.created_at)
        except ValueError:
            raise InvalidArgumentError("created_at must be ISO-8601") from None
    return AgentFeedback(media_id=msg.media_id or None, goal=msg.goal or None,
                         mode=mode_from_proto(msg.mode).value, corrected_tasks=list(msg.corrected_tasks),
                         created_at=created_at, **values)


def feedback_to_proto(feedback: AgentFeedback) -> pb.AgentFeedback:
    msg = pb.AgentFeedback(media_id=feedback.media_id or 0, goal=feedback.goal or "",
                           mode=mode_to_proto(AnalysisMode.from_nullable(feedback.mode)),
                           corrected_tasks=[t for t in feedback.corrected_tasks or [] if t is not None],
                           created_at="" if feedback.created_at is None else format_instant(feedback.created_at))
    for name in _FEEDBACK_OPTIONAL:
        value = getattr(feedback, name)
        if value is not None:
            setattr(msg, name, value)
    return msg


# ---------- telemetry seed / free-form JSON ----------
def seed_from_proto(request: pb.RunRequest) -> dict | None:
    """Gateway telemetry seed in the shape AgentTelemetry.apply_seed consumes."""
    if not request.HasField("telemetry_seed"):
        return None
    seed = request.telemetry_seed
    return {
        "traceId": seed.trace_id or None,
        "stages": {name: {"durationMs": t.duration_ms, "success": t.success} for name, t in seed.stages.items()},
        "counters": dict(seed.counters),
    }


def to_struct(value: dict | None) -> struct_pb2.Struct:
    """JSON object -> Struct. Numbers become doubles, as in any JSON transport."""
    struct = struct_pb2.Struct()
    if value:
        struct.update(value)
    return struct
